pytensor/link/jax/ops.py

```diff
@@ -0,0 +1,424 @@
+        gz = tree_map(
+            lambda g, primal: jnp.broadcast_to(g, jnp.shape(primal)),
+            gz,
+            primals,
+        )
+        if len(y0) == 1:
+            return vjp_fn(gz)[0]
+        else:
+            return tuple(vjp_fn(gz))
+
+    return vjp_sol_op_jax
+
+
+def _partition_jaxfunc(jaxfunc, static_vars, func_vars):
+    """Partition the jax function into static and non-static variables.
+
+    Returns a function that accepts only non-static variables and returns the non-static
+    variables. The returned static variables are stored in a dictionary and returned,
+    to allow the referencing after creating the function
+
+    Additionally wrapped functions saved in func_vars are regenerated with
+    vars["vars_from_func"] as input, to allow the transformation of the variables.
+    """
+    static_out_dic = {"out": None}
+
+    def jaxfunc_partitioned(vars):
+        vars, vars_from_func = vars["vars"], vars["vars_from_func"]
+        func_vars_evaled = tree_map(
+            lambda x, y: x.get_func_with_vars(y), func_vars, vars_from_func
+        )
+        args, kwargs = eqx.combine(
+            vars, static_vars, func_vars_evaled, is_leaf=callable
+        )
+
+        out = jaxfunc(*args, **kwargs)
+        outvars, static_out = eqx.partition(out, eqx.is_array, is_leaf=callable)
+        static_out_dic["out"] = static_out
+        return outvars
+
+    return jaxfunc_partitioned, static_out_dic
+
+
+### Construct the function that accepts flat inputs and returns flat outputs.
+def _flatten_func(jaxfunc, vars_treedef):
+    def func_flattened(vars_flat):
+        vars = tree_unflatten(vars_treedef, vars_flat)
+        outvars = jaxfunc(vars)
+        outvars_flat, _ = tree_flatten(outvars)
+        return _normalize_flat_output(outvars_flat)
+
+    return func_flattened
+
+
+def _normalize_flat_output(output):
+    if len(output) > 1:
+        return tuple(
+            output
+        )  # Transform to tuple because jax makes a difference between
+        # tuple and list and not pytensor
+    else:
+        return output[0]
+
+
+def _return_pytensor_ops_classes(name):
+    class SolOp(Op):
+        def __init__(
+            self,
+            input_treedef,
+            output_treeedef,
+            input_types,
+            output_types,
+            jitted_sol_op_jax,
+            jitted_vjp_sol_op_jax,
+        ):
+            self.vjp_sol_op = None
+            self.input_treedef = input_treedef
+            self.output_treedef = output_treeedef
+            self.input_types = input_types
+            self.output_types = output_types
+            self.jitted_sol_op_jax = jitted_sol_op_jax
+            self.jitted_vjp_sol_op_jax = jitted_vjp_sol_op_jax
+
+        def make_node(self, *inputs):
+            self.num_inputs = len(inputs)
+
+            # Define our output variables
+            outputs = [pt.as_tensor_variable(type()) for type in self.output_types]
```

Comment:
Would it be possible to use jax machinery to infer the output types from the input types? Can we created TraceDArrays (or whatever they're called) as pass them through the function?

Reply:
Scrap that, JAX doesn't let you trace arrays without unknown shape

Reply:
I trace the shape through the JAX function in line 119 of the file. It won't work for unknown shape. But if one specifies the shape at the beginning of a graph, i.e. `x = pm.Normal("x", shape=(3,))`, and it loses static shape information afterwards, for instance because of a pt.cumsum, line 99 (`pytensor.compile.builders.infer_shape`) will be able to infer the shape. But that is a good comment, I will raise an error if `pytensor.compile.builders.infer_shape` isn't able to infer the shape. I think it makes sense to only use this wrapper if the shape information is known.

Reply:
Okay, I see a point where it will lead to problems: If one has as an input a `x = pm.Data("x", shape=(None,), value= np.array([0., 0]))`. In the first run, it will work, as `pytensor.compile.builders.infer_shape` will infer the shape as (2,), but if one changes with x.set_value(np.array([0., 0, 0]))` the shape of `x`, it will lead to an error, as the Pytensor Op has been created with an explicit shape. I could simply add a parameter to  `as_jax_op` to simply force all output shapes to `None`, then it should work.

Reply:
I will write more tests, then it will be clearer what I mean

Reply:
Yeah, we can't use shape unless it's static. Ideally it shouldn't fail for unknown shapes, but then the user has to tell user the output types. 

We can allow the user to specify a make_node callable? That way it can be made to work with different dtypes/ndims if the jax function handles those fine